### ISSUE TITLE
"Not a git repository" Error

### ISSUE THREAD

**Issue by AndrewCSQ** [sodapopcan/vim-twiggy#30]
Trying to use what looks like a fantastic plugin, running into some issues

## System Information
- OS: Linux (Fedora 30)
- Vim: Neovim (0.3.6)
- Terminal: Kitty and Konsole
- Plugin Manager: Plug

## Replication
- cd to home directory
- mkdir Test2 (any random directory)
- cd Test2
- git init (to initialize it as a git repository)
- run :Twiggy, returns "Not a git repository", the Twiggy panel says 'merge in progress', when aborted, says "/bin/bash: -c: line 0: unexpected EOF while looking for matching `'' " and "syntax error: unexpected end of file"

I have removed all other plugins and theming (keeping only fugitive and vim-twiggy) to eliminate other possible causes.

Issue also persists when I try to clone my own GitHub repos (freshly made and existing projects)

**Comment by sodapopcan** [comment]
Oh my!  What version of git are you using?

**Comment by AndrewCSQ** [comment]
2.21 it's strange because I recall it worked for just the briefest moment (and in that brief moment was some pretty awesome stuff) and then it went back to not working. I'm kind of new to (neo)vim (but not to programming as a whole), so I'm sorry that I'm taking up your time with a debugging request

**Comment by sodapopcan** [comment]
Oh, not at all!  'tis the nature of open source :)  I'm actually working on twiggy today so it's especially no problem.

There have been some plumbing changes recently that have caused a few issues for some folks.  Unfortunately, I dev on a mac so I can't reproduce your set up exactly but I'll see what I can do.

**Comment by sodapopcan** [comment]
Would you mind cloning twiggy, checkout `20cdca4` of this copy then point Plug to it and restart nvim?  This was the last commit before I started to make some deeper changes.  It doesn't really seem like it should cause this type of error, but it would be nice to rule it out.

Also, do you have Vim 8 on your machine?  Is it the same problem if you do?

Also, thank you for your kind words!  So sorry it's not working for you.

**Comment by AndrewCSQ** [comment]
Under neovim, the issue persists even with the given commit. Either "Not a git repository" or the same EOF errors as above.

Under vim (both master and the commit mentioned) I get:
Error detected while processing twiggy#Branch[14]..<SNR>28_Render[39]..<SNR>28_no_commits[1]..<SNR>28_git_cmd[6]..<SNR>28_system:
line 11:
E484: Can't open file /tmp/vvDbu4G/0

I suspect they stem from the same issue.

**Comment by sodapopcan** [comment]
Interesting!  Thank you so much for taking the time to do that.

Do you mind letting me know what the output of this vim command is:

```viml
:echo systemlist('git rev-list -n 1 --all | wc -l')
```

There are some warnings regarding the use of `system()` in the docs and I have noticed that many of tpope's plugins have a custom `shellescape()` function and I've never known the reason.  When I blindly copy said function from vim-dispatch and apply it to all commands, I actually get an error similar to yours!  I'm gonna dig in a little more.  It seems like it's a system and/or shell thing.

**Comment by AndrewCSQ** [comment]
The vim output is ['1']
I've heard the vim issue mentioned by a different issue on your repo, setting shell=/bin/bash has no effect (as expected on a Fedora system where it is the default shell anyway.

**Comment by sodapopcan** [comment]
Yeah, I'm thinking it's more of a system issue than a shell issue.  I don't think setting the shell makes a difference anymore (and I don't think it affects linux users at all... from memory, it was to get macOS to use the right version of git or something).

`:h system()` has a bit of info.  I'm going to try some stuff out and push a branch you can try.

Thanks for your patience with this!

**Comment by AndrewCSQ** [comment]
I can confirm that the same error somehow persists on my mac partition (running macOS Mojave, with the latest versions of everything installed through brew)

**Comment by sodapopcan** [comment]
That is so strange.  I've been having a very hard time reproducing it.  All I can come close to is calling some version of `shellescape` and then I get a sort of similar error.  I had to stop hacking on it the other day but hopefully get back to it in the next few days.  Thanks for the update!

**Comment by Five52** [comment]
Hi,
I have the same issue when I am on an empty tab. I think it may come from a recent commit from Fugitive, which doesn't work properly on empty tabs anymore. The same issue appears when I use `:Gstatus` on an empty tab, but `:G` works fine.

**Comment by mredaelli** [comment]
I have the same issue witn `nvim`, and I also get `['1']` as output to that command.

**Comment by jmolero** [comment]
I also had the `merge in progress` problem. In my case it was because I had the project inside a path with a folder with a space in its name. There were no problems after I moved it to another path without folders with spaces.

**Comment by arnarthor** [comment]
I had this issue, I tried running `:Twiggy master` to switch branches, which worked, then I tried `:Twiggy` again and everything worked fine after that 🤷‍♂

**Comment by jericdeleon** [comment]
Same problem on latest Arch box on Vagrant (https://app.vagrantup.com/archlinux/boxes/archlinux)

**Comment by j-sv** [comment]
I've had the `Not a git repository` issue, but seems unlikely it's the same cause. For me, it seems [this commit](https://github.com/tpope/vim-fugitive/commit/49287b6e874cf31bf3fc2e239be21b096ff4cc3e) broke something. Specifically, this line:

`autocmd VimEnter * if empty(expand('<amatch>'))|call FugitiveDetect(getcwd())|endif`

If I add that to my `.vimrc`, `:Twiggy` seem to work just fine again.

**Comment by faisal-shah** [comment]
+1 to this issue. Have the same problem.

If I open a file within the repo, and then call :Twiggy, no problem. But with an empty buffer, this problem persists.

**Comment by languitar** [comment]
Has there ever been a solution for this one. I'm hit by this now too.

**Comment by thomas-ah** [comment]
Hiya! I absolutely love this plugin, but am also getting this issue since yesterday. Welp! :(

Is there any way to debug this maybe? Using Neovim 0.5.1 on Arch btw.

**Comment by sodapopcan** [comment]
Hey all,

Apologies for the inattention here.

Does this happen even when you edit a file that is tracked by git?  I can reproduce if I open vim in a repo and no file loaded in the current buffer.

I'll take some time today and/or tomorrow to see if I can fix it.  Twiggy was my first ever vim-plugin and is full of tech-debt and very odd decisions, so some serious refactoring is in order.

**Comment by cd-a** [comment]
I have the same issue. Open nvim, open Twiggy, same message.

Then I open fugitive `:Git` once, and close it. Then Twiggy opens properly without complaining.

**Comment by frenkel** [comment]
Same as @dahmc here. I'm on Fedora using gvim.

**Comment by sinetodev** [comment]
The workaround that @j-sv did seems to work in feel cases when starts inside a git project. But if I open a file of this project the error persists. Same for @dahmc advice.

https://user-images.githubusercontent.com/61889013/162455041-0b5cc598-36f2-4164-9c39-88f83b7ab370.mp4

Any solution?

**Comment by devenv** [comment]
`Lazy` config for lazies to make `Twiggy` work at the start:
```
{
"tpope/vim-fugitive",
event = "VeryLazy",
  config = function()
    vim.cmd("call FugitiveDetect(getcwd())")
  end
},
```
Seems like this needs to be run after each `cd` (or session load), so hook appropriately.

**Comment by kristoferfannar** [comment]
> `Lazy` config for lazies to make `Twiggy` work at the start:
> 
> ```
> {
> "tpope/vim-fugitive",
> event = "VeryLazy",
>   config = function()
>     vim.cmd("call FugitiveDetect(getcwd())")
>   end
> },
> ```
> 
> Seems like this needs to be run after each `cd` (or session load), so hook appropriately.

I had the same error on nvim 0.9, macOS 13, Twiggy gave me the "not a git repo" error until after I had used a fugitive command within the session. 

Adding this into my lazy config allows me to use `Twiggy` before having to do that just fine.

Edit:
Actually, that's not the case - I'm back to nvim since october, but still get this error even with the hackfix above...

**Comment by minusf** [comment]
This is still an issue on neovim 0.11.4 and latest fugitive...

Unless I do `:call FugitiveDetect(getcwd())|Twiggy` it's always "not in a git repo" error message.

Perhaps an important addition: I use tabs for files. Opening a new tab inside the same git repo triggers the issue and I can reproduce in every time. Running the detect command in every new tab makes Twiggy work...

**Comment by sodapopcan** [comment]
Does this still happen after editing a file?  I get it if I open vim without editing a file but otherwise it always works.

I do need to give Twiggy some love, it's been a while.
